### ISSUE TITLE
Error: "Reply could not be parsed: invalid character '<'" when processing metrics

### ISSUE THREAD

**Issue by metmajer** [hawkular/hawkular-openshift-agent#165]
I am using the agent configuration from https://github.com/openshift/origin-metrics/blob/master/hawkular-agent/hawkular-openshift-agent-configmap.yaml and deployed HOSA into the `default` project to gather metrics from all pods.

However, metrics cannot be obtained due to the following metrics processing errors:

```
…
W0415 20:30:54.905664       1 metrics_storage.go:101] Failed to determine if metric definition [pod/6ba6bb7d-221a-11e7-a78d-12d904fd13ae/custom/process_resident_memory_bytes] of type [gauge] in tenant [default] exists. err=Hawkular returned status code 403, error message: Reply could not be parsed: invalid character '<' looking for beginning of value
W0415 20:30:54.906817       1 metrics_storage.go:101] Failed to determine if metric definition [pod/6ba6bb7d-221a-11e7-a78d-12d904fd13ae/custom/process_start_time_seconds] of type [gauge] in tenant [default] exists. err=Hawkular returned status code 403, error message: Reply could not be parsed: invalid character '<' looking for beginning of value
W0415 20:30:54.907929       1 metrics_storage.go:101] Failed to determine if metric definition [pod/6ba6bb7d-221a-11e7-a78d-12d904fd13ae/custom/process_virtual_memory_bytes] of type [gauge] in tenant [default] exists. err=Hawkular returned status code 403, error message: Reply could not be parsed: invalid character '<' looking for beginning of value
I0415 20:31:24.879745       1 prometheus_metrics_collector.go:103] DEBUG: Told to collect [22] Prometheus metrics from [http://172.17.0.2:8080/metrics]
W0415 20:31:24.915114       1 metrics_storage.go:149] Failed to store metrics. err=Hawkular returned status code 403, error message: Reply could not be parsed: invalid character '<' looking for beginning of value
2017/04/15 20:31:34 Unsolicited response received on idle HTTP channel starting with "HTTP/1.0 408 Request Time-out\r\nCache-Control: no-cache\r\nConnection: close\r\nContent-Type: text/html\r\n\r\n<html><body><h1>408 Request Time-out</h1>\nYour browser didn't send a complete request in time.\n</body></html>\n"; err=<nil>
I0415 20:31:54.879802       1 prometheus_metrics_collector.go:103] DEBUG: Told to collect [22] Prometheus metrics from [http://172.17.0.2:8080/metrics]
…
```

Here's the entire agent configuration for the sake of completeness (the only changes have been made to the `hawkular_server` directive):

```
id: hawkular-openshift-agent
kind: ConfigMap
apiVersion: v1
name: Hawkular OpenShift Agent Configuration
metadata:
  name: hawkular-openshift-agent-configuration
  labels:
    metrics-infra: agent
data:
  config.yaml: |
    hawkular_server:
      url: "https://metrics-openshift-infra.1.2.3.4.nip.io"
      tls:
        skip_certificate_validation: true
    kubernetes:
      tenant: ${POD:namespace_name}
    collector:
      minimum_collection_interval: 10s
      default_collection_interval: 30s
      metric_id_prefix: pod/${POD:uid}/custom/
      tags:
        metric_name: ${METRIC:name}
        description: ${METRIC:description}
        units: ${METRIC:units}
        namespace_id: ${POD:namespace_uid}
        namespace_name: ${POD:namespace_name}
        node_name: ${POD:node_name}
        pod_id: ${POD:uid}
        pod_ip: ${POD:ip}
        pod_name: ${POD:name}
        pod_namespace: ${POD:namespace_name}
        hostname: ${POD:hostname}
        host_ip: ${POD:host_ip}
        labels: ${POD:labels}
        type: pod
        collector: hawkular_openshift_agent
        custom_metric: true
  hawkular-openshift-agent: |
    endpoints:
    - type: prometheus
      protocol: "http"
      port: 8080
      path: /metrics
      collection_interval: 30s
      metrics:
      - name: hawkular_openshift_agent_metric_data_points_collected_total
      - name: process_cpu_seconds_total
      - name: go_goroutines
      - name: go_memstats_alloc_bytes
      - name: go_memstats_gc_sys_bytes
      - name: go_memstats_heap_alloc_bytes
      - name: go_memstats_heap_idle_bytes
      - name: go_memstats_heap_inuse_bytes
      - name: go_memstats_heap_objects
      - name: go_memstats_heap_released_bytes
      - name: go_memstats_heap_sys_bytes
      - name: go_memstats_last_gc_time_seconds
      - name: go_memstats_next_gc_bytes
      - name: go_memstats_other_sys_bytes
      - name: go_memstats_stack_inuse_bytes
      - name: go_memstats_stack_sys_bytes
      - name: go_memstats_sys_bytes
      - name: process_max_fds
      - name: process_open_fds
      - name: process_resident_memory_bytes
      - name: process_start_time_seconds
      - name: process_virtual_memory_bytes
```

**Comment by jmazzitelli** [comment]
Looks like the Origin Metrics server is not able to process requests from HOSA.

```
Hawkular returned status code 403, error message: Reply could not be parsed:
   invalid character '<' looking for beginning of value
Unsolicited response received on idle HTTP channel starting with "HTTP/1.0 408
  Request Time-out\r\nCache-Control: no-cache\r\nConnection: close\r\nContent-Type: 
  text/html\r\n\r\n<html><body><h1>408 Request Time-out</h1>\nYour browser didn't
  send a complete request in time.\n</body></html>\n"; err=<nil>
```
No idea why the metrics server is returning this. Never saw this before.

@mwringe @jsanda you ever see this error code/message from Hawkular Metrics?

**Comment by jmazzitelli** [comment]
BTW: if you want more debug log messages, in your openshift template yaml for HOSA, where you see this: https://github.com/openshift/origin-metrics/blob/master/hawkular-agent/hawkular-openshift-agent.yaml#L76-L77

```
command:
- "/opt/hawkular/hawkular-openshift-agent"
- "-config"
- "/hawkular-openshift-agent-configuration/config.yaml"
- "-v"
- "4"
```

You can change that "-v 4" option to "-v 5" and get lots of trace output. Not sure if it will help, but it will show lots more debug info. "-v 4" turns on debug messages which is less noisy but will still output some debug messages.

**Comment by metmajer** [comment]
@jmazzitelli thanks for your hint with regards to `-v`. Unfortunately, it didn't bring any additional insights into the particular issue. Looking forward to @mwringe's and @jsanda's toughts. Thanks!

**Comment by mwringe** [comment]
The 403 error is 'forbidden' and if you are getting that from the Hawkular Metrics server, it likely means you are not getting the right password from the Hawkular Metrics secret.

Can you please let us know what images you are using for the metric components and for HOSA?

**Comment by metmajer** [comment]
@mwringe, here you go:

### origin-metrics

```
openshift/origin-metrics-cassandra                          v3.6.0-alpha.0      0b26a8a2f16f
openshift/origin-metrics-hawkular-metrics                   v3.6.0-alpha.0      c5ffcd045782
openshift/origin-metrics-heapster                           v3.6.0-alpha.0      066918c75f9c
openshift/origin-metrics-deployer                           v3.6.0-alpha.0      e14d7be77c30
```

### hosa

```
hawkular/hawkular-openshift-agent                           1.0.0.Final         7982ad7e5add
```

**Comment by jmazzitelli** [comment]
OOOOhhhh.. @metmajer  that is a very old HOSA. That might explain some things :)

Here you can see where we publish our HOSA versions:

https://hub.docker.com/r/hawkular/hawkular-openshift-agent/tags/

I suggest pulling the "latest" - which is 1.4.1.Final.

**Comment by metmajer** [comment]
Oh dear :-) I'll immediately give `latest` a try and update the open tickets accordingly. Would be good though to, once in a while, update https://github.com/openshift/origin-metrics/blob/master/hawkular-agent/hawkular-openshift-agent.yaml#L10-L11.

**Comment by jmazzitelli** [comment]
@metmajer yeah, we need to do a better job coordinating docs and configs across HOSA and Origin Metrics repos.

FWIW: I am getting an error with the latest builds and latest HOSA and I believe it is due to changes in the latest Origin Metrics - it is what @mwringe mentioned above. I'm going to get with matt and find out what changes need to be made due to these OpenShift Origin changes - but if you use an older version of OpenShift, the latest HOSA should work.

**Comment by jmazzitelli** [comment]
quick smoke test shows this appears to fix the problem (in agent's configuration configmap):

```
hawkular_server:
  ca_cert_file: secret:openshift-infra/hawkular-metrics-certs/tls.truststore.crt
```

**Comment by metmajer** [comment]
@jmazzitelli Nice! Just tried your suggestion, but I am getting the following error:

```
E0419 14:36:16.225489       1 metrics_storage.go:201] Error trying to get Secret named [hawkular-metrics-certs] from namespace [openshift-infra]: err=secrets "hawkular-metrics-certs" not found
```

Therefore, I renamed your suggestion to what came closest: there's a secret named `hawkular-metrics-certificate` in `openshift-infra`, which contains `hawkular-metrics-ca.certificate` and `hawkular-metrics.certificate`.

After changing the ConfigMap and redeploying HOSA, I am receiving the following issue:

```
I0419 14:43:14.977714       1 emitter_server.go:48] Agent emitter will emit metrics
I0419 14:43:14.977724       1 emitter_server.go:64] Agent emitter will NOT emit status
I0419 14:43:14.977728       1 emitter_server.go:70] Agent emitter will provide a health probe
I0419 14:43:14.977733       1 emitter_server.go:96] Agent will start the emitter endpoint at [:8080]
E0419 14:43:15.003322       1 metrics_storage.go:217] Could not get the requested secret. This may mean the secret has not yet been created yet or the agent does not have the required permission to access the secret. Will attempt again every 5 seconds for the next 5 minutes
```

I am still using `v3.6.0-alpha.0`. If this is where things finally break, I'll move on to `1.5.0.rc0`. Suggestions?

**Comment by metmajer** [comment]
Seems like the reason why HOSA stops is because neither `secret:openshift-infra/hawkular-metrics-certs/tls.truststore.crt` nor `secret:openshift-infra/hawkular-metrics-certificate/tls.truststore.crt` exists.

**Comment by jmazzitelli** [comment]
@metmajer i got those when I ran my latest OpenShift build. I'm on this call - should be done in about an hour. Let's get on the phone afterwards and we can talk through this. Hop on #hawkular (freenode) in about an hour and let's coordinate. I have a feeling its just version mismatching and changes that went into latest OpenShift

**Comment by jmazzitelli** [comment]
@metmajer  I see you are in Austria (correct?) its probably late over there. when you are available, hop on #hawkular (freenode) and we can discuss when you are free. should be better than a phone call

**Comment by jpkrohling** [comment]
@metmajer I'm on the same timezone as you, so, feel free to ping me.

**Comment by metmajer** [comment]
Thanks @jmazzitelli and @jpkrohling for your great help. I've been able to get Joloka metrics fed into hawkular-metrics using HOSA using 3.6.0.alpha.0. I'll provide an update tomorrow!

**Comment by metmajer** [comment]
The problem was entirely resolved using the latest `1.4.1-Final` tag of the HOSA Docker image. Sorry for the fuzz, but I suggest you either keep the agent template in sync or at least change the documentation to use a `oc process ... -p DOCKER_IMAGE=latest` as `1.0.0-Final` is the hard-coded default, currently.

**Comment by akram** [comment]
Hi all,
it looks like that hawkular installation on openshift 3.6 has 2 differences. To make it work:
edit the config map
```
ca_cert_file: secret:openshift-infra/hawkular-metrics-certificate/hawkular-metrics.certificate
```
the secret name and the file name are different
